cmake/FindPhysX.cmake

```diff
@@ -1,77 +1,128 @@
-        ${PHYSX_PHYSICS_INCLUDE_DIR}
-        ${PHYSX_LOADER_INCLUDE_DIR}
+macro(_find_physx_library VAR_NAME)
+    find_library(${VAR_NAME}_LIBRARY_RELEASE
+                 NAMES ${_${VAR_NAME}_LIBRARY_NAMES}
+                 PATHS ${_PHYSX_RELEASE_PREFIX}
+                 PATH_SUFFIXES lib
+                 NO_CMAKE_PATH
+    )
+    find_library(${VAR_NAME}_LIBRARY_DEBUG
+                 NAMES ${_${VAR_NAME}_LIBRARY_NAMES}
+                 PATHS ${_PHYSX_DEBUG_PREFIX}
+                 PATH_SUFFIXES lib
+                 NO_CMAKE_PATH
     )
+    select_library_configurations(${VAR_NAME})
+endmacro()
+
+_find_physx_library(PHYSX_COMMON)
+_find_physx_library(PHYSX_PHYSICS)
+_find_physx_library(PHYSX_FOUNDATION)
+_find_physx_library(PHYSX_COOKING)
+_find_physx_library(PHYSX_EXTENSIONS)
+_find_physx_library(PHYSX_CHARACTER)
+_find_physx_library(PHYSX_PVD)
+
+find_path(PHYSX_INCLUDE_DIR NAMES PxPhysicsAPI.h PATH_SUFFIXES physx)
+mark_as_advanced(PHYSX_INCLUDE_DIR)
+
+# PhysX has two include directories, but vcpkg merges them together. In case of a manual build,
+# check and compensate for the lack of merging here.
+if(PHYSX_INCLUDE_DIR)
+    if(NOT EXISTS ${PHYSX_INCLUDE_DIR}/foundation/Px.h)
+        find_path(PHYSX_FOUNDATION_INCLUDE_DIR NAMES foundation/Px.h
+                  PATH_SUFFIXES pxshared/include)
+        mark_as_advanced(PHYSX_FOUNDATION_INCLUDE_DIR)
+    else()
+        set(PHYSX_FOUNDATION_INCLUDE_DIR ${PHYSX_INCLUDE_DIR} CACHE PATH "Path to a file.")
+    endif()
 endif()
 
-#2.7.3 is the earliest version that can be downloaded from nVidia's archive.
-find_library(PHYSX_COOKING_LIBRARY NAMES NxCooking
-             PATH_SUFFIXES "PhysX/v2.7.3"
-             PATHS "${PHYSX_SDK_PATH}/lib/win32"
-)
-find_library(PHYSX_CHARACTER_LIBRARY NAMES NxCharacter
-             PATH_SUFFIXES "PhysX/v2.7.3"
-             PATHS "${PHYSX_SDK_PATH}/lib/win32"
-)
-find_library(PHYSX_EXTENSIONS_LIBRARY NAMES NxExtensions
-             PATH_SUFFIXES "PhysX/v2.7.3"
-             PATHS "${PHYSX_SDK_PATH}/lib/win32"
-)
-find_library(PHYSX_LOADER_LIBRARY NAMES PhysXLoader
-             PATHS "${PHYSX_SDK_PATH}/lib/win32"
+find_package_handle_standard_args(PhysX
+                                  REQUIRED_VARS PHYSX_COMMON_LIBRARY
+                                                PHYSX_PHYSICS_LIBRARY
+                                                PHYSX_FOUNDATION_LIBRARY
+                                                PHYSX_COOKING_LIBRARY
+                                                PHYSX_EXTENSIONS_LIBRARY
+                                                PHYSX_CHARACTER_LIBRARY
+                                                PHYSX_PVD_LIBRARY
+                                                PHYSX_INCLUDE_DIR
+                                                PHYSX_FOUNDATION_INCLUDE_DIR
+                                  REASON_FAILURE_MESSAGE "Be sure that PhysX 4.1 is available."
 )
 
 set(PHYSX_LIBRARIES
```

Comment:
To successfully link on Linux, the order here must be exactly:

```cmake
set(PHYSX_LIBRARIES
    ${PHYSX_CHARACTER_LIBRARY}
    ${PHYSX_EXTENSIONS_LIBRARY}
    ${PHYSX_PHYSICS_LIBRARY}
    ${PHYSX_PVD_LIBRARY}
    ${PHYSX_COOKING_LIBRARY}
    ${PHYSX_COMMON_LIBRARY}
    ${PHYSX_FOUNDATION_LIBRARY}
)
```

Reply:
Weird, but ok.